lib/html.rb

```diff
@@ -16,39 +16,127 @@
 
 class Html
-  def self.escape_javascript(object)
-    object.gsub(/([\\'"])/, "\\\\\\1")
+  module Helpers
+    def scope_help(scope)
+      text = File.read(File.join(Machinery::ROOT, "plugins", "#{scope}/#{scope}.md"))
+      Kramdown::Document.new(text).to_html
+    end
+
+    def diff_to_object(diff)
+      diff = Machinery.scrub(diff)
+      lines = diff.lines[2..-1]
+      diff_object = {
+        file: diff[/--- a(.*)/, 1],
+        additions: lines.select { |l| l.start_with?("+") }.length,
+        deletions: lines.select { |l| l.start_with?("-") }.length
+      }
+
+      original_line_number = 0
+      new_line_number = 0
+      diff_object[:lines] = lines.map do |line|
+        line = ERB::Util.html_escape(line.chomp).
+          gsub("\\", "&#92;").
+          gsub("\t", "&nbsp;"*8)
+        case line
+        when /^@.*/
+          entry = {
+            type: "header",
+            content: line
+          }
+          original_line_number = line[/-(\d+)/, 1].to_i
+          new_line_number = line[/\+(\d+)/, 1].to_i
+        when /^ .*/, ""
+          entry = {
+            type: "common",
+            new_line_number: new_line_number,
+            original_line_number: original_line_number,
+            content: line[1..-1]
+          }
+          new_line_number += 1
+          original_line_number += 1
+        when /^\+.*/
+          entry = {
+            type: "addition",
+            new_line_number: new_line_number,
+            content: line[1..-1]
+          }
+          new_line_number += 1
+        when /^\-.*/
+          entry = {
+            type: "deletion",
+            original_line_number: original_line_number,
+            content: line[1..-1]
+          }
+          original_line_number += 1
+        end
+
+        entry
+      end
+
+      diff_object
+    end
   end
 
-  def self.generate(description)
-    template = Haml::Engine.new(
-      File.read(File.join(Machinery::ROOT, "html", "index.html.haml"))
-    )
-    target = description.store.description_path(description.name)
-
-    diffs_dir = description.scope_file_store("analyze/config_file_diffs").path
-    if description.config_files && diffs_dir
-      # Enrich description with the config file diffs
-      description.config_files.files.each do |file|
-        path = File.join(diffs_dir, file.name + ".diff")
-        file.diff = diff_to_object(File.read(path)) if File.exists?(path)
+  # this is required for the #generate_comparison method which renders a HAML template manually with the local binding,
+  # so it expects the helper methods to be available in Html. It can be removed oncen the comparison was move to the
+  # webserver approach as well
+  extend Helpers
+
+  # Creates a new thread running a sinatra webserver which serves the local system descriptions
+  # The Thread object is returned so that the caller can `.join` it until it's finished.
+  def self.run_server(opts)
+    Thread.new do
+      require 'sinatra/base'
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.